File: backend/app/api/routes/leveling.py
"""Leveling configuration API routes."""
from fastapi import APIRouter
from typing import Dict, List, Any, Optional
from pydantic import BaseModel, Field

from ...models.leveling_config import (
    get_complete_level_config,
    generate_level_progression,
    get_unlocked_gimmicks,
    get_phase_for_level,
    calculate_level_difficulty,
    PROFESSIONAL_GIMMICK_UNLOCK,
    PHASE_CONFIGS,
    SAWTOOTH_PATTERN_10,
    DEFAULT_PROFESSIONAL_UNLOCK_LEVELS,
)


router = APIRouter(prefix="/api/leveling", tags=["leveling"])


class LevelConfigRequest(BaseModel):
    """단일 레벨 설정 요청"""
    level_number: int = Field(..., ge=1, description="레벨 번호 (1-based)")
    available_gimmicks: Optional[List[str]] = Field(
        default=None,
        description="사용 가능한 기믹 풀 (None = 모든 기믹)"
    )
    use_sawtooth: bool = Field(
        default=True,
        description="톱니바퀴 난이도 패턴 적용"
    )


class LevelProgressionRequest(BaseModel):
    """레벨 진행 계획 요청"""
    start_level: int = Field(default=1, ge=1, description="시작 레벨")
    count: int = Field(default=10, ge=1, le=200, description="생성할 레벨 수")
    available_gimmicks: Optional[List[str]] = Field(
        default=None,
        description="사용 가능한 기믹 풀"
    )
    use_sawtooth: bool = Field(
        default=True,
        description="톱니바퀴 패턴 적용"
    )


class GimmickUnlockInfo(BaseModel):
    """기믹 언락 정보"""
    gimmick: str
    unlock_level: int
    practice_levels: int
    integration_start: int
    difficulty_weight: float
    description: str


@router.get("/config")
async def get_leveling_config() -> Dict[str, Any]:
    """
    프로 레벨링 시스템 전체 설정 반환

    Returns:
        - gimmick_unlocks: 기믹별 언락 정보
        - phase_configs: 단계별 설정
        - sawtooth_pattern: 톱니바퀴 난이도 패턴
        - unlock_levels: 간단한 언락 레벨 맵
    """
    return {
        "gimmick_unlocks": {
            name: {
                "gimmick": config.gimmick,
                "unlock_level": config.unlock_level,
                "practice_levels": config.practice_levels,
                "integration_start": config.integration_start,
                "difficulty_weight": config.difficulty_weight,
                "description": config.description,
            }
            for name, config in PROFESSIONAL_GIMMICK_UNLOCK.items()
        },
        "phase_configs": {
            phase.value: {
                "phase": config.phase.value,
                "level_range": config.level_range,
                "min_tile_types": config.min_tile_types,
                "max_tile_types": config.max_tile_types,
                "min_layers": config.min_layers,
                "max_layers": config.max_layers,
                "max_gimmick_types": config.max_gimmick_types,
                "base_difficulty": config.base_difficulty,
                "difficulty_increment": config.difficulty_increment,
            }
            for phase, config in PHASE_CONFIGS.items()
        },
        "sawtooth_pattern": SAWTOOTH_PATTERN_10,
        "unlock_levels": DEFAULT_PROFESSIONAL_UNLOCK_LEVELS,
    }


@router.post("/level-config")
async def get_single_level_config(request: LevelConfigRequest) -> Dict[str, Any]:
    """
    단일 레벨의 권장 설정 반환

    Returns:
        - level_number: 레벨 번호
        - phase: 현재 단계
        - difficulty: 목표 난이도
        - tile_types_count: 권장 타일 종류 수
        - layer_count: 권장 레이어 수
        - gimmicks: 권장 기믹
        - is_tutorial_level: 튜토리얼 레벨 여부
        - is_boss_level: 보스 레벨 여부
        - tutorial_gimmick: 튜토리얼 기믹 (있는 경우)
    """
    return get_complete_level_config(
        request.level_number,
        request.available_gimmicks,
        request.use_sawtooth
    )


@router.post("/progression")
async def get_level_progression(request: LevelProgressionRequest) -> List[Dict[str, Any]]:
    """
    연속 레벨들의 진행 계획 반환

    여러 레벨을 한 번에 생성할 때 사용합니다.
    각 레벨의 난이도, 기믹, 타일 종류 등을 자동으로 계산합니다.

    Returns:
        레벨별 설정 리스트
    """
    return generate_level_progression(
        request.start_level,
        request.count,
        request.available_gimmicks,
        request.use_sawtooth
    )


@router.get("/unlocked-gimmicks/{level_number}")
async def get_unlocked_gimmicks_at_level(level_number: int) -> Dict[str, Any]:
    """
    특정 레벨에서 사용 가능한 기믹 목록 반환

    Args:
        level_number: 레벨 번호

    Returns:
        - level_number: 레벨 번호
        - unlocked_gimmicks: 언락된 기믹 목록
        - total_available: 사용 가능한 총 기믹 수
        - phase: 현재 단계
    """
    unlocked = get_unlocked_gimmicks(level_number)
    phase = get_phase_for_level(level_number)

    return {
        "level_number": level_number,
        "unlocked_gimmicks": unlocked,
        "total_available": len(unlocked),
        "phase": phase.value,
    }


@router.get("/difficulty-curve")
async def get_difficulty_curve(
    start_level: int = 1,
    count: int = 50,
    use_sawtooth: bool = True
) -> Dict[str, Any]:
    """
    난이도 곡선 데이터 반환 (차트 그리기용)

    Args:
        start_level: 시작 레벨
        count: 레벨 수
        use_sawtooth: 톱니바퀴 패턴 적용

    Returns:
        - levels: 레벨 번호 배열
        - difficulties: 난이도 배열
        - phases: 단계 배열
        - boss_levels: 보스 레벨 인덱스 배열
        - tutorial_levels: 튜토리얼 레벨 정보
    """
    levels = []
    difficulties = []
    phases = []
    boss_levels = []
    tutorial_levels = []

    for i in range(count):
        level_num = start_level + i
        difficulty = calculate_level_difficulty(level_num, use_sawtooth)
        phase = get_phase_for_level(level_num)

        levels.append(level_num)
        difficulties.append(round(difficulty, 3))
        phases.append(phase.value)

        if level_num % 10 == 0:
            boss_levels.append(level_num)

        # 튜토리얼 레벨 체크
        for config in PROFESSIONAL_GIMMICK_UNLOCK.values():
            if config.unlock_level == level_num:
                tutorial_levels.append({
                    "level": level_num,
                    "gimmick": config.gimmick,
                    "description": config.description,
                })

    return {
        "levels": levels,
        "difficulties": difficulties,
        "phases": phases,
        "boss_levels": boss_levels,
        "tutorial_levels": tutorial_levels,
    }

File: frontend/src/api/leveling.ts
/**
 * 프로 레벨링 시스템 API
 *
 * 유명 타일 게임(Tile Buster, Tile Explorer 등)의
 * 레벨링 패턴을 분석하여 구현한 시스템입니다.
 */

const API_BASE = '/api';

// =========================================================
// 타입 정의
// =========================================================

/** 레벨 진행 단계 */
export type LevelPhase =
  | 'tutorial'      // 1-10: 순수 매칭 학습
  | 'basic'         // 11-50: 단일 기믹 학습
  | 'intermediate'  // 51-100: 2기믹 조합
  | 'advanced'      // 101-150: 3기믹 조합
  | 'expert'        // 151-200: 4기믹 조합
  | 'master';       // 201+: 최고 난이도

/** 기믹 도입 단계 */
export type GimmickIntroPhase =
  | 'tutorial'      // 첫 등장 레벨
  | 'practice'      // 연습 레벨
  | 'integration'   // 이전 기믹과 통합
  | 'mastery';      // 숙달 단계

/** 기믹 언락 설정 */
export interface GimmickUnlockConfig {
  gimmick: string;
  unlock_level: number;
  practice_levels: number;
  integration_start: number;
  difficulty_weight: number;
  description: string;
}

/** 단계별 설정 */
export interface PhaseConfig {
  phase: LevelPhase;
  level_range: [number, number];
  min_tile_types: number;
  max_tile_types: number;
  min_layers: number;
  max_layers: number;
  max_gimmick_types: number;
  base_difficulty: number;
  difficulty_increment: number;
}

/** 전체 레벨링 설정 */
export interface LevelingConfig {
  gimmick_unlocks: Record<string, GimmickUnlockConfig>;
  phase_configs: Record<LevelPhase, PhaseConfig>;
  sawtooth_pattern: number[];
  unlock_levels: Record<string, number>;
}

/** 단일 레벨 설정 */
export interface LevelConfig {
  level_number: number;
  phase: LevelPhase;
  difficulty: number;
  tile_types_count: number;
  layer_count: number;
  gimmicks: string[];
  gimmick_intro_phase: GimmickIntroPhase;
  is_tutorial_level: boolean;
  is_boss_level: boolean;
  tutorial_gimmick: string | null;
}

/** 난이도 곡선 데이터 */
export interface DifficultyCurveData {
  levels: number[];
  difficulties: number[];
  phases: LevelPhase[];
  boss_levels: number[];
  tutorial_levels: Array<{
    level: number;
    gimmick: string;
    description: string;
  }>;
}

// =========================================================
// API 함수
// =========================================================

/**
 * 전체 레벨링 설정 가져오기
 */
export async function getLevelingConfig(): Promise<LevelingConfig> {
  const response = await fetch(`${API_BASE}/leveling/config`);
  if (!response.ok) {
    throw new Error(`Failed to get leveling config: ${response.statusText}`);
  }
  return response.json();
}

/**
 * 단일 레벨 설정 가져오기
 */
export async function getLevelConfig(
  levelNumber: number,
  availableGimmicks?: string[],
  useSawtooth: boolean = true
): Promise<LevelConfig> {
  const response = await fetch(`${API_BASE}/leveling/level-config`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      level_number: levelNumber,
      available_gimmicks: availableGimmicks,
      use_sawtooth: useSawtooth,
    }),
  });
  if (!response.ok) {
    throw new Error(`Failed to get level config: ${response.statusText}`);
  }
  return response.json();
}

/**
 * 레벨 진행 계획 가져오기 (여러 레벨)
 */
export async function getLevelProgression(
  startLevel: number,
  count: number,
  availableGimmicks?: string[],
  useSawtooth: boolean = true
): Promise<LevelConfig[]> {
  const response = await fetch(`${API_BASE}/leveling/progression`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      start_level: startLevel,
      count: count,
      available_gimmicks: availableGimmicks,
      use_sawtooth: useSawtooth,
    }),
  });
  if (!response.ok) {
    throw new Error(`Failed to get level progression: ${response.statusText}`);
  }
  return response.json();
}

/**
 * 특정 레벨에서 사용 가능한 기믹 목록
 */
export async function getUnlockedGimmicks(levelNumber: number): Promise<{
  level_number: number;
  unlocked_gimmicks: string[];
  total_available: number;
  phase: LevelPhase;
}> {
  const response = await fetch(`${API_BASE}/leveling/unlocked-gimmicks/${levelNumber}`);
  if (!response.ok) {
    throw new Error(`Failed to get unlocked gimmicks: ${response.statusText}`);
  }
  return response.json();
}

/**
 * 난이도 곡선 데이터 가져오기 (차트용)
 */
export async function getDifficultyCurve(
  startLevel: number = 1,
  count: number = 50,
  useSawtooth: boolean = true
): Promise<DifficultyCurveData> {
  const params = new URLSearchParams({
    start_level: startLevel.toString(),
    count: count.toString(),
    use_sawtooth: useSawtooth.toString(),
  });
  const response = await fetch(`${API_BASE}/leveling/difficulty-curve?${params}`);
  if (!response.ok) {
    throw new Error(`Failed to get difficulty curve: ${response.statusText}`);
  }
  return response.json();
}

// =========================================================
// 프론트엔드 전용 프로페셔널 언락 레벨 (API 없이 사용)
// =========================================================

/**
 * 프로페셔널 기믹 언락 레벨
 * 유명 타일 게임 패턴 분석 기반
 */
export const PROFESSIONAL_UNLOCK_LEVELS: Record<string, number> = {
  chain: 11,
  ice: 21,
  frog: 36,
  grass: 51,
  link: 66,
  bomb: 81,
  curtain: 96,
  teleport: 111,
  unknown: 126,
  craft: 141,
  stack: 156,
};

/**
 * 단계별 레벨 범위
 */
export const PHASE_LEVEL_RANGES: Record<LevelPhase, [number, number]> = {
  tutorial: [1, 10],
  basic: [11, 50],
  intermediate: [51, 100],
  advanced: [101, 150],
  expert: [151, 200],
  master: [201, 9999],
};

/**
 * 톱니바퀴 난이도 패턴 (10레벨 단위)
 */
export const SAWTOOTH_PATTERN: number[] = [
  0.0,   // 레벨 1: 새 시작
  0.1,   // 레벨 2: 쉬움
  0.2,   // 레벨 3: 약간 증가
  0.35,  // 레벨 4: 중간 진입
  0.45,  // 레벨 5: 중간
  0.55,  // 레벨 6: 중간 위
  0.7,   // 레벨 7: 어려움 시작
  0.85,  // 레벨 8: 어려움
  0.75,  // 레벨 9: 보스 전 휴식
  1.0,   // 레벨 10: 보스
];

/**
 * 레벨 번호에 해당하는 단계 반환
 */
export function getPhaseForLevel(levelNumber: number): LevelPhase {
  for (const [phase, [start, end]] of Object.entries(PHASE_LEVEL_RANGES)) {
    if (levelNumber >= start && levelNumber <= end) {
      return phase as LevelPhase;
    }
  }
  return 'master';
}

/**
 * 레벨 번호에서 언락된 기믹 목록 반환
 */
export function getUnlockedGimmicksLocal(levelNumber: number): string[] {
  return Object.entries(PROFESSIONAL_UNLOCK_LEVELS)
    .filter(([_, unlockLevel]) => unlockLevel <= levelNumber)
    .map(([gimmick]) => gimmick);
}

/**
 * 레벨이 튜토리얼 레벨인지 확인
 */
export function isTutorialLevel(levelNumber: number): string | null {
  for (const [gimmick, unlockLevel] of Object.entries(PROFESSIONAL_UNLOCK_LEVELS)) {
    if (unlockLevel === levelNumber) {
      return gimmick;
    }
  }
  return null;
}

/**
 * 레벨이 보스 레벨인지 확인
 */
export function isBossLevel(levelNumber: number): boolean {
  return levelNumber % 10 === 0;
}
